### ISSUE TITLE
CME Equities Trading Hours

### ISSUE THREAD

**Issue by glossner** [rsheftel/pandas_market_calendars#182]
As discussed in the [original thread](https://github.com/rsheftel/pandas_market_calendars/issues/180#issuecomment-1048128077), Sunday is not considered a normal trading day. For now, let's keep that the working definition and if appropriate I'll come back to it later and open it as a separate enhancement.

CME lists the following for[ trading hours:](https://www.cmegroup.com/trading-hours.html)
- Sunday 5:00 p.m. - Friday 5:45 p.m. CT with no reporting Monday - Thursday 5:45 p.m. - 6:00 p.m. CT

This is coded the exchange_calendar_cme.py as:
```
    regular_market_times = {
        "market_open": ((None, time(17), -1),), # offset by -1 day
        "market_close": ((None, time(16)),),
        "break_start": ((None, time(15,15)),),
        "break_end": ((None, time(15,30)),)
    }
```
This doesn't match the CME link. 

Not being from this domain, I've also just spent an embarrassing amount of time trying to find the actual equities trading times. There is another set of times at the same CME link called Globex. It gives this table for the S&P E-minis.

Product Name | Sunday 13 Feb 2022 | Monday 14 Feb 2022 | Tuesday 15 Feb 2022
-- | -- | -- | --
ES E-mini S&P 500 Futures | TRADE DATE: Mon 14 Feb 16:00 preopen 17:00 open | TRADE DATE: Tue 15 Feb 16:45 preopen 17:00 open | TRADE DATE: Wed 16 Feb 16:45 preopen 17:00 open

I interpret this as break_start should be 16:45 and break_end should be 17:00 except for Sunday. However, let's table different regular_market_times on different days for a future enhancement issue. 

For now, if anyone has pointers as to what should be the equities market trading hours it would be appreciated. Thanks!

**Comment by rsheftel** [comment]
When looking at information on the CME web site be very careful to note what time zone they are using. Since the CME is in Central Time, often their listed times are in that time zone. So a break starting at 16:45 (aka: 4:45) on their web site is referring to 16:45 CST.

In regards to the main point, it is very possible that the CME items in the code currently are stale or not up to date, so treat everything in code as suspect until proven correct.

You will see times listed as "pit", that is the old world when things only traded in the pit. These are no longer important for us. The GLOBEX is the name of their electronic matching engine and these are the hours that the market trades to today given that 99-100% of the volume is electroinic.

**Comment by muttermachine** [comment]
Firstly, thank you for all the effort you and the other contributors put into this library.  

I am not an expert on the nuances, but have a fair bit of experience with Futures and the CME.  I am sure I am technically wrong on a few points, but conceptually the CME has evolved into a collection of exchanges, which all share a common technology platform for electronic trading "Globex".  Similarly, they have "Clearport" for the settlement side.  The "Globex" hours you linked to should be considered the correct source IMO, but I find it much easier to read the product pages (eg [S&P](https://www.cmegroup.com/markets/equities/sp/e-mini-sandp500.contractSpecs.html)).

For the "e-minis" as the S&P futures are commonly called, they trade (Chicago time):
*  Sunday 17:00 - Monday 16:00
* Monday 17:00 - Tuesday 16:00
* Tues 17:00 - Wed 16:00
* Wed 17:00 - Thur 16:00
* Thursday 17:00 - Friday 16:00

One challenge with the CME, as noted in the original thread, is that trading hours vary by asset class/sector/product, and they can be quite different on holidays.  It should be noted that even "equities" have different hours at holidays, due to the wide variety of listings.  I am not sure what I would recommend, other than perhaps, for now, keeping it simple and sticking with the most liquid futures when deciding what to do.

**Comment by rsheftel** [comment]
@muttermachine thanks for the input.

I agree that the way to handle this is to treat each commodity class (Equities, Bonds, etc) as a separate "exchange" in the code. And for some classes, like energies, each individual market (Oil, Natural Gas) is also unique, so the crude oil contract needs to be its own "exchange". Its a bit bending the terminology but those in futures will definitely understand.

**Comment by glossner** [comment]
@muttermachine  Thank you very much for the GLOBEX S&P E-mini link. I can now find all the trading times for all products.

I will be using the GLOBEX trading times unless I see additional comments to this post.

**Comment by muttermachine** [comment]
I would be happy to divide and conquer and push up some code for the different CME calendars.  I would prefer to start from a finalised example, so perhaps after the equity one is built we can split up the workload a bit?

**Comment by glossner** [comment]
I've been looking at different open/close times of different products.

CME classifies products by market. For example, Agricultural Products include Grains, Oilseeds, Livestock, etc.

Underneath Grains are Corn, Chicago Wheat, and KC Wheat. These all have the same trading time but different from S&P.

Underneath Livestock are Live Cattle, Feeder Cattle, Lean Hog, and Pork Cutout. These all have the same trading time but different from both S&P and Grains.

I plan to design the CME Exchange Calendar Hierarchy to match the CME classification. e.g.,  

CMEBaseExchangeCalendar -> CMEAgricultureExchangeCalendar -> CMELivestockExchangeCalendar

Then use aliases for ["CME_Live_Cattle", "CME_Feeder_Cattle", etc.]

I looked at a sampling of Holidays. It appears this will  work and minimize code duplication.

Anybody have alternative design suggestions? If not I'll work on a full example with S&P and Livestock.

**Comment by Stryder-Git** [comment]
@glossner, I think your design suggestion makes a lot of sense, and I don't have a better one. 

@muttermachine, thanks for your help! I also agree with:
> I would be happy to divide and conquer and push up some code for the different CME calendars. I would prefer to start from a finalised example, so perhaps after the equity one is built we can split up the workload a bit?

How about you just let us know when your full example is ready, @glossner, and then we can divide up the rest of the calendars? 

On a side note, I find the 'different regular_market_times on different days'-problem an interesting one. I will try to work out a couple of ideas on how to integrate this in the next few weeks and keep you posted!

**Comment by glossner** [comment]
@muttermachine @Stryder-Git  I pushed the Livestock holidays to my fork. Turns out that was a bad example to choose because they trade M-F 8:30-13:05 which means they follow normal holiday full close patterns. I'll pick a different CME product that trades Sundays but I won't get to it until next weekend earliest. 

Another thought, if we're really going to cover all the CME products we might want to break up the files by Market (e.g. exchange_calendar_cme_agriculture, exchange_calendar_cme_equity_index, etc.). That would also help with divide and conquer.

**Comment by glossner** [comment]
@MADChittum  yes, I can put that back in and have it end 6/28/2021. Thanks for catching that.

I will also check the other Markets.

**Comment by muttermachine** [comment]
Re: national days of mourning, I found a couple of indications that equity and interest rate products observed them:

https://www.cmegroup.com/media-room/press-releases/2006/12/29/cme_announces_markethoursofoperationinhonorofnationaldayofmourni.html

https://www.cmegroup.com/media-room/press-releases/2018/12/02/cme_group_u_s_equityinterestratemarketstoclosefornationaldayofmo.html


It also reads as if all of Globex was closed in 2004 "CME's trading floors will be closed on Friday, June 11 for the National Day of Mourning, while GLOBEX markets will close for trading at 8 a.m." - note also the extension by 30 minutes on June 10th:
http://investor.cmegroup.com/news-releases/news-release-details/cme-extend-june-10-floor-trading-session-interest-rate-fx

**Comment by Stryder-Git** [comment]
@glossner, @muttermachine 

> I'll pick a different CME product that trades Sundays but I won't get to it until next weekend earliest

I have actually been hit hard by the flu, so definitely take your time, my productivity is stalled anyway.  But I will try to have a closer look at your fork during the weekend to see how I can help out.

**Comment by glossner** [comment]
@Stryder-Git  sorry to hear about that. I recovered from Covid a couple months ago. It was not fun. Get well!

@muttermachine I updated my fork with the naming and calendar factorization by Markets. I modified your class names to exactly match the way CME lists them on their website. I'm very sorry if that clobbers anything you've done since your last commit. Going forward this should also allow us to work on different markets without stepping on toes.

**Comment by Stryder-Git** [comment]
Hi @glossner and @muttermachine, 

I feel like myself again and managed to catch up on some work. I also had a look through your repo, @glossner. I see that Agriculture (i.e. Livestock), and FX are being worked on. Do you think I should just get started with Energy, implementing it in the same style, or do you have a different suggestion on how I could help out?

Thanks!

**Comment by glossner** [comment]
@Stryder-Git  Very glad t hear you are back to health.

I'm guessing the equities calendar (S&P E-mini) is a needed one.
https://www.cmegroup.com/markets/equities.html

I didn't check but it likely follows the NYSE pattern.

**Comment by Stryder-Git** [comment]
@glossner, thanks! 

I will have a look at the equities section then!

**Comment by glossner** [comment]
@Stryder-Git  If you find a better class factorization or have other input, I'm not adverse to changing the hierarchy.

I'm likely 2-3 weeks until I complete the Agricultural markets.

Thanks for the link to the script.

**Comment by Stryder-Git** [comment]
@rsheftel, @glossner, @muttermachine

How do you think we should proceed?

I see two general areas that would need work

* Finishing the rest of the 'exchanges'
	If I am not mistaken we have done FX, Agriculture and Equities, which ones do you think we should tackle next, or did one of you already start on a different one?


* More flexible market times
	Honestly, I don't remember exactly where this was needed. It also seems related to #193. If you could point me to any similar cases where this was needed/beneficial, I could see if I can find a good way to update the framework.

On a side note, if you have any clue what is causing Travis to fail in #189, I would be very happy to hear it.

Thanks!

**Comment by glossner** [comment]
@Stryder-Git Adding a break in the middle of the trading day was the other time improvement. These are commented as "NOT IMPLEMENTED" in exchange_calendar_nyse.py. Note that some events have multiple short closes per day. Also, in 1 case, trading was extended 15 minutes. 

I have a slight preference to continue working on the CME globex calendars. I have no preference in which order the remaining ones get done. If we choose this we should perhaps open a new thread and just write what one we're working on to avoid overlap.

Thanks.

**Comment by rsheftel** [comment]
I agree that completing more exchanges adds the most value to people. I suggest the order of:

1. Fixed Income
2. Energies
3. Metals

**Comment by Stryder-Git** [comment]
I'm fine taking Fixed Income. I'll also try to whip up a proposal to improve the time handling. I am expecting this to be about a month, and I will just open a PR for it.

**Comment by glossner** [comment]
I'll take Energies.

**Comment by glossner** [comment]
Turns out Energies and Metals follow the same trading times and holiday schedule so I'll take both of those.

**Comment by Stryder-Git** [comment]
Regarding #197 and the following steps:

Some special opens are missing (see comments containing "5am" in tests) and regular market_open/_close seem to have changed at some point. I am going to leave this for now and implement it when I am working on #198, since that may cause a few more changes.

Additionally:
  * Do you agree with the aliases and name for the Fixed Income Calendar?
  * Can the file exchange_calendar_cme.py be removed completely?

**Comment by rsheftel** [comment]
I would add to the aliases:

bond_futures
interest_rate_futures

**Comment by rsheftel** [comment]
@Stryder-Git Is this complete in #210 ? Should we close this?

**Comment by glossner** [comment]
I'm closing this. I believe it is complete. We can reopen later if needed. Thanks.
